lib/srv/db/opensearch/engine.go

```diff
@@ -0,0 +1,331 @@
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	//nolint:bodyclose // resp will be closed in sendResponse().
+	resp, err := tr.RoundTrip(signedReq)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	return trace.Wrap(e.sendResponse(resp))
+}
+
+func (e *Engine) getTransport(ctx context.Context) (*http.Transport, error) {
+	tr, err := defaults.Transport()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	tlsConfig, err := e.Auth.GetTLSConfig(ctx, e.sessionCtx)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	tr.TLSClientConfig = tlsConfig
+	return tr, nil
+}
+
+func (e *Engine) getSignedRequest(reqCopy *http.Request) (*http.Request, error) {
+	roleArn, err := libaws.BuildRoleARN(e.sessionCtx.DatabaseUser, e.sessionCtx.Database.GetAWS().Region, e.sessionCtx.Database.GetAWS().AccountID)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	signCtx := &libaws.SigningCtx{
+		SigningName:   opensearchservice.EndpointsID,
+		SigningRegion: e.sessionCtx.Database.GetAWS().Region,
+		Expiry:        e.sessionCtx.Identity.Expires,
+		SessionName:   e.sessionCtx.Identity.Username,
+		AWSRoleArn:    roleArn,
+		AWSExternalID: e.sessionCtx.Database.GetAWS().ExternalID,
+	}
+
+	signedReq, err := e.signingSvc.SignRequest(e.Context, reqCopy, signCtx)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return signedReq, nil
+}
+
+func (e *Engine) rewriteRequest(ctx context.Context, req *http.Request) (*http.Request, error) {
+	payload, err := utils.GetAndReplaceRequestBody(req)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	reqCopy := req.Clone(ctx)
+	reqCopy.RequestURI = ""
+	reqCopy.Body = io.NopCloser(bytes.NewReader(payload))
+
+	// Connection is hop-by-hop header, drop.
+	reqCopy.Header.Del("Connection")
+
+	// force HTTPS, set host URL.
+	reqCopy.URL.Scheme = "https"
+	reqCopy.URL.Host = e.sessionCtx.Database.GetURI()
+	reqCopy.Host = e.sessionCtx.Database.GetURI()
+
+	return reqCopy, nil
+}
+
+// emitAuditEvent writes the request and response to audit stream.
+func (e *Engine) emitAuditEvent(req *http.Request) {
+	// Try to read the body and JSON unmarshal it.
+	// If this fails, we still want to emit the rest of the event info; the request event Body is nullable,
+	// so it's ok if body is nil here.
+	body, err := libaws.UnmarshalRequestBody(req)
+	if err != nil {
+		e.Log.WithError(err).Warn("Failed to read request body as JSON, omitting the body from the audit event.")
+	}
+
+	ev := &apievents.OpenSearchRequest{
+		Metadata: common.MakeEventMetadata(e.sessionCtx,
+			events.DatabaseSessionOpenSearchRequestEvent,
+			events.OpenSearchRequestCode),
+		UserMetadata:     common.MakeUserMetadata(e.sessionCtx),
+		SessionMetadata:  common.MakeSessionMetadata(e.sessionCtx),
+		DatabaseMetadata: common.MakeDatabaseMetadata(e.sessionCtx),
+		Method:           req.Method,
+		Path:             req.URL.Path,
+		RawQuery:         req.URL.RawQuery,
+		Body:             body,
+	}
+
+	e.Audit.EmitEvent(req.Context(), ev)
```

Comment:
I think that we should youse internal process context instead of request context that can be canceled aany time by a client leading to interruption of `EmitEvent` flow 
```suggestion
	e.Audit.EmitEvent(e.Context, ev)
```

Reply:
Fixed.